### PR TITLE
[#159990511] integrate application insights

### PR DESCRIPTION
**Pull request by gunzip**
This PR does nothing by itself. It adds an ApplicationInsights winston transport that must be set up
before calling any log* method (see [here](https://github.com/teamdigitale/italia-backend/pull/293/files#diff-d746dc18af77137dbd3942db2b6e368aR9)).

To make this work one must set a new environment variable `APPINSIGHTS_INSTRUMENTATIONKEY` (into the kubernetes configuration for the deployment and/or using kubectl). Ask the Azure administrator for the value of this variable.

### CODE REVIEW

**Comment by digitalcitizenship** [comment]
## Affected stories

  * 🌟 [#159990511](https://www.pivotaltracker.com/story/show/159990511): Come AMS voglio poter ricevere degli alert in caso di errori del backend


New dependencies added: <a href='https://www.npmjs.com/package/applicationinsights'><code>applicationinsights</code></a>.


<h2><a href="https://github.com/Microsoft/ApplicationInsights-node.js#readme">applicationinsights</a></h2>
<p>Author: Unknown</p>
<p>Description: Microsoft Application Insights module for Node.js</p>
<p>Homepage: <a href="https://github.com/Microsoft/ApplicationInsights-node.js#readme">https://github.com/Microsoft/ApplicationInsights-node.js#readme</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>about 4 years ago</td></tr><tr><td>Last Updated</td><td>3 months ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>47</td></tr><tr><td>Direct Dependencies</td><td><a href='https://www.npmjs.com/package/diagnostic-channel'><code>diagnostic-channel</code></a>, <a href='https://www.npmjs.com/package/diagnostic-channel-publishers'><code>diagnostic-channel-publishers</code></a> and <a href='https://www.npmjs.com/package/zone.js'><code>zone.js</code></a></td></tr><tr><td>Keywords</td><td>exception monitoring, request monitoring, performance monitoring, application insights, microsoft, azure, tracing, telemetry, analytics and apm</td></tr>
</table>
This README is too long to show.

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by cloudify** [comment]
Does this send all logs to Appinsights? right now we're logging "secret" stuff like authentication tokens and SAML assertions, is there a way to filter out those?

**Comment by gunzip** [comment]
an option is to use log.debug() to log secrets and disable that in "production". another one is to log a json object instead of a simple string and filter out some value like [this](https://github.com/winstonjs/winston#filtering-info-objects)

**Comment by cloudify** [comment]
@lussoluca what do you think?

File: src/utils/applicationinsights.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+export function wrapCustomTelemetryClient(
+  isTracingDisabled: boolean,
+  client: ApplicationInsights.TelemetryClient
+): (
+  params?: ITelemetryParams,
+  commonProperties?: Record<string, string>
+) => ApplicationInsights.TelemetryClient {
+  if (isTracingDisabled) {
+    // this won't disable manual calls to trackEvent / trackDependency
+    ApplicationInsights.Configuration.setAutoCollectConsole(false)
```
Comment (cloudify) [inline src/utils/applicationinsights.ts]: not sure it makes sense to have this side effect here, this gets called every time we create a new client and will also change the appinsights configuration globally

Comment (cloudify) [reply src/utils/applicationinsights.ts]: also, I don't see any call to this function in the code? where does it gets used?

Comment (gunzip) [reply src/utils/applicationinsights.ts]: This code is now in io-functions-commons here: https://github.com/teamdigitale/io-functions-commons/blob/master/src/utils/application_insights.ts

` this gets called every time we create a new client` <-- the applicationinsights client (like the other azure clients) should be a singleton so `wrapCustomTelemetryClient` should be called once in the bootstrap phase (its purpose is to globally configure the (singleton) appinsights client).

Comment (cloudify) [reply src/utils/applicationinsights.ts]: ops I didn't notice this was the old PR :) pls check the new one https://github.com/teamdigitale/io-backend/pull/503

**Comment by cloudify** [comment]
Superseded by #503

**Pull request closed without merging**